ublox/src/ubx_packets/packets.rs

```diff
@@ -1939,111 +1941,123 @@ struct RxmRtcm {
+        size_fn = data_len,
+        is_valid = EsfMeasDataIter::is_valid,
+        may_fail,
+    )]
+    data: [u8; 0],
     #[ubx(
-        map_type = EsfMeasInfo,
-        from = EsfMeasInfo::new,
-        flatten,
-        flat_fields = ["flags", "u16", "id", "u16", "data", "U32Iter"]
+        map_type = Option<u32>,
+        from = EsfMeas::calib_tag,
+        size_fn = calib_tag_len,
     )]
-    //, "calib_tag", "Option<u32>"
-    info: [u8; 0],
+    calib_tag: [u8; 0],
 }
 
-#[derive(serde::Serialize, Debug)]
-pub struct EsfMeasInfo<'a> {
-    pub flags: u16,
-    pub id: u16,
-    pub data: U32Iter<'a>,
-    pub calib_tag: Option<u32>,
+impl EsfMeas {
+    fn calib_tag(bytes: &[u8]) -> Option<u32> {
+        bytes.try_into().ok().map(u32::from_le_bytes)
+    }
 }
 
-#[ubx_iter]
-pub struct U32Iter<'a>(core::slice::ChunksExact<'a, u8>);
+impl<'a> EsfMeasRef<'a> {
+    fn data_len(&self) -> usize {
+        ((self.flags() >> 11 & 0x1f) as usize) * 4
+    }
+
+    fn calib_tag_len(&self) -> usize {
+        if self.flags() & 0x8 != 0 {
+            4
+        } else {
+            0
+        }
+    }
+}
 
-impl<'a> U32Iter<'a> {
+#[derive(Debug, serde::Serialize)]
+pub struct EsfMeasData {
+    pub data_type: u8,
+    pub data_field: u32,
+}
+
+#[derive(Debug, Clone)]
+pub struct EsfMeasDataIter<'a>(core::slice::ChunksExact<'a, u8>);
+
+impl<'a> EsfMeasDataIter<'a> {
     fn new(bytes: &'a [u8]) -> Self {
-        U32Iter(bytes.chunks_exact(4))
+        Self(bytes.chunks_exact(4))
     }
 
     fn is_valid(bytes: &'a [u8]) -> bool {
         bytes.len() % 4 == 0
     }
 }
 
-impl<'a> core::iter::Iterator for U32Iter<'a> {
-    type Item = u32;
+impl<'a> core::iter::Iterator for EsfMeasDataIter<'a> {
+    type Item = EsfMeasData;
 
     fn next(&mut self) -> Option<Self::Item> {
-        self.0
-            .next()
-            .map(|bytes| u32::from_le_bytes(bytes.try_into().unwrap()))
-    }
-}
-
-impl<'a> EsfMeasInfo<'a> {
-    fn new(bytes: &'a [u8]) -> Self {
-        let flags = u16::from_le_bytes(bytes[0..2].try_into().unwrap());
-        let id = u16::from_le_bytes(bytes[2..4].try_into().unwrap());
-        let num_meas = usize::from(flags >> 11 & 0x1F);
-        let data = U32Iter::new(&bytes[4..num_meas * 4]);
-        let calib_tag = if flags & 0x8 != 0 {
-            let slice = bytes[num_meas * 4 + 4..num_meas * 4 + 8]
-                .try_into()
-                .unwrap();
-            Some(u32::from_le_bytes(slice))
-        } else {
-            None
-        };
-
-        Self {
-            flags,
-            id,
-            data,
-            calib_tag,
-        }
+        let data = self.0.next()?.try_into().map(u32::from_le_bytes).unwrap();
+        Some(EsfMeasData {
+            data_type: ((data & 0x3F000000) >> 24).try_into().unwrap(),
+            data_field: data & 0xFFFFFF,
+        })
     }
 }
 
 #[ubx_packet_recv]
 #[ubx(class = 0x10, id = 0x03, max_payload_len = 1240)]
```

Comment:
also, since we can just use size_fn, would it be better to make it so `max_payload_len` isn't hard coded to 1240? - this was a number couldn't figure out the meaning to since it factorizes as 
31 * 5 * 2 * 2 * 2
no clue where or how 1240 came from, or what the 5 means, 31 maybe 2^5 - 1

Reply:
I'm not sure, I figured it was just the max payload length from the spec but I don't see one. It might be arbitrary, having a max length makes no_std easier (without a resizable vec handling an unknown max size is hard). I don't think we can replace it because the values not known till runtime for the variable length message. Or do you mean change it so that it's a 1240-FIXED_BYTES % item_len == 0 or something?

Reply:
Ya as far as I can tell max length should actually be 2^16?

Reply:
hmmmmm it just has 1240 as default which when i first edited this repo was really odd number,
i dont think it does much with this number tho so maybe not a big deal but wasn't sure if that was the true upper bound

maybe we just put as 2^16=65536 as well then (wasn't sure if original author had some reason for that)